alchemiscale_fah/compute/client.py

```diff
@@ -0,0 +1,608 @@
+
+        self.as_url = as_url.geturl()
+        self.ws_url = ws_url.geturl()
+        self.ws_ip_addr = socket.gethostbyname(ws_url.hostname)
+
+        # self.certificate = (
+        #    self.read_certificate(certificate_file) if certificate_file else None
+        # )
+
+        # if key_file is None:
+        #    self.key = self.create_key()
+        # else:
+        #    self.key = self.read_key(key_file)
+
+        self.certificate_file = (
+            Path(certificate_file).absolute() if certificate_file is not None else None
+        )
+        self.key_file = Path(key_file).absolute() if key_file is not None else None
+        self.csr_file = Path(csr_file).absolute() if csr_file is not None else None
+
+        self.cert = (self.certificate_file, self.key_file)
+        self.verify = verify
+
+    @staticmethod
+    def read_key(key_file):
+        with open(key_file, "rb") as f:
+            pem = f.read()
+
+        return serialization.load_pem_private_key(pem, None, default_backend())
+
+    @staticmethod
+    def read_certificate(certificate_file):
+        with open(certificate_file, "rb") as f:
+            pem = f.read()
+
+        return x509.load_pem_x509_certificate(pem, default_backend())
+
+    @classmethod
+    def create_key(cls):
+        return rsa.generate_private_key(
+            backend=default_backend(), public_exponent=65537, key_size=4096
+        )
+
+    @classmethod
+    def write_key(cls, key, key_file):
+        pem = key.private_bytes(
+            encoding=serialization.Encoding.PEM,
+            format=serialization.PrivateFormat.TraditionalOpenSSL,
+            encryption_algorithm=serialization.NoEncryption(),
+        )
+
+        with open(key_file, "wb") as f:
+            f.write(pem)
+
+    @classmethod
+    def generate_csr(cls, key, csr_file, email):
+        """Generate certificate signing request (CSR) using private key.
+
+        It is necessary to create a CSR and present this to an AS in order to
+        receive a valid certificate. The CSR will be written in PEM format.
+
+        Parameters
+        ----------
+        key
+            RSA private key, as returned by `create_key` or `read_key`.
+        csr_file
+            Path in which to place CSR generated by this method.
+        email
+            Email address used to authenticate with the AS.
+
+        """
+        cn = x509.NameAttribute(NameOID.COMMON_NAME, email)
+        csr = x509.CertificateSigningRequestBuilder()
+        csr = csr.subject_name(x509.Name([cn]))
+        csr = csr.sign(key, hashes.SHA256())
+
+        with open(csr_file, "wb") as f:
+            f.write(csr.public_bytes(serialization.Encoding.PEM))
+
+    @staticmethod
+    def read_csr_pem(csr_file):
+        with open(csr_file, "r") as f:
+            pem = f.read()
+
+        return pem
+
+    def _check_status(self, r):
+        if r.status_code != 200:
+            raise Exception("Request failed with %d: %s" % (r.status_code, r.text))
```

Comment:
If you want to catch and then reraise somthing like the following would work, otherwise can just use raw raise_for_status

```python
try: 
   r.raise_for_status()
except HTTPError as e:
    raise Exception(f"Request failed with {r.status_code}, {r.text}") from e

```

Reply:
Switched to raw `raise_for_status`!